src/NuGet.Core/NuGet.Protocol/Model/ServiceIndexEntry.cs

```diff
@@ -26,7 +28,7 @@ public class ServiceIndexEntry
         /// </summary>
         public SemanticVersion ClientVersion { get; }
 
-        public ServiceIndexEntry(Uri serviceUri, string serviceType, SemanticVersion clientVersion)
+        internal ServiceIndexEntry(Uri serviceUri, string serviceType, SemanticVersion clientVersion, PackageSource packageSource)
```

Comment:
If you just raise the event from the ServiceIndexResourceV3 you don't need all these changes. 
Makes the PR footprint a lot smaller.